### ISSUE TITLE
shuffleNet onnx转ncnn不成功， 由于gather op

### ISSUE THREAD

**Issue by luoduo21** [Tencent/ncnn#1425]
shuffle onnx gather op用onnx-simplifier去除不了。

**Comment by Note-Liu** [comment]
楼主有尝试过直接shuffle net的caffe转成ncnn吗？

**Comment by luoduo21** [comment]
caffe我没试过,应该没问题吧,pytorch有点问题,pytorch torchvisions实现的shuffleNet没有gather op,但是官方实现就有gather op

**Comment by nihui** [comment]
https://github.com/megvii-model/ShuffleNet-Series
你用的是这里的 shufflenet 实现吗？

**Comment by luoduo21** [comment]
> https://github.com/megvii-model/ShuffleNet-Series
> 你用的是这里的 shufflenet 实现吗？
是的,最后我用slice代替gather解决了

**Comment by sunnymoon155** [comment]
@luoduo21 请问具体是怎么解决的呢？我目前也遇到了这个问题，如何使用slice代替gather呢？

**Comment by luoduo21** [comment]
@sunnymoon155    
我只改了这个
    def channel_shuffle(self, x):
        batchsize, num_channels, height, width = x.data.size()
        channels_per_group = num_channels // 2

        # reshape
        x = x.view(batchsize, channels_per_group,
                   2, height, width)

        x = torch.transpose(x, 1, 2).contiguous()

        # flatten
        x = x.view(batchsize, -1, height, width)
        return x.chunk(2, dim=1)

**Comment by nihui** [comment]
最新的 onnx2ncnn 可以无痛转换 pytorch vision 里的 shufflenet_v2 模型了

**Comment by sunnymoon155** [comment]
@nihui @luoduo21 谢谢两位大佬！
经过修改已经可以转ncnn了，可是使用int8量化还是出了问题。。
./ncnn2table --param shuffle-sim.param --bin shufflenetv2-sim.bin --images imgs/ --output shufflenetv2-sim.table --mean 127.5,127.5,127.5 --norm 0.0078125,0.0078125,0.0078125 --size 224,224 --thread 2
--- ncnn post training quantization tool --- 07:49:29 Nov 7 2019
param = 'shufflenetv2-sim.param'
bin = 'shufflenetv2-sim.bin'
images = 'imgs/'
output = 'shufflenetv2-sim.table'
mean = '127.5,127.5,127.5'
norm = '0.0078125,0.0078125,0.0078125'
size = '224,224'
thread = '2'
====> Quantize the parameters.
====> Quantize the activation.
====> step 1 : find the max value.
[1] 2429 segmentation fault (core dumped) /home/luban/ncnn-master/build/tools/quantize/ncnn2table --param --bin

**Comment by sunnymoon155** [comment]
使用大佬提供的官方shufflenetv2可以成功转换int8了哈哈，谢谢

**Comment by Note-Liu** [comment]
> 使用大佬提供的官方shufflenetv2可以成功转换int8了哈哈，谢谢
你好，能加你交流一下吗

**Comment by Note-Liu** [comment]
> @sunnymoon155
> 我只改了这个
> def channel_shuffle(self, x):
> batchsize, num_channels, height, width = x.data.size()
> channels_per_group = num_channels // 2
>     # reshape
>     x = x.view(batchsize, channels_per_group,
>                2, height, width)
> 
>     x = torch.transpose(x, 1, 2).contiguous()
> 
>     # flatten
>     x = x.view(batchsize, -1, height, width)
>     return x.chunk(2, dim=1)
大佬，你的pytorch怎么转的onnx?

**Comment by sunnymoon155** [comment]
@Note-Liu 我换了官方的shufflenetv2模型就可以转了
https://pytorch.org/hub/pytorch_vision_shufflenet_v2/

**Comment by DuckJ** [comment]
@nihui 大佬，请问ncnn examples中[https://github.com/Tencent/ncnn/blob/master/examples/shufflenetv2.cpp#L35](https://github.com/Tencent/ncnn/blob/master/examples/shufflenetv2.cpp#L35)，这个shufflenet v2通过onnx可以转成功ncnn么，我看ncnn例子用的是这个模型

**Comment by nihui** [comment]
最新的 onnx2ncnn 支持了 pytorch megvii shufflenet 的 channel shuffle 和 split channel

**Comment by lgcy** [comment]
> 最新的 onnx2ncnn 支持了 pytorch megvii shufflenet 的 channel shuffle 和 split channel

试了一下貌似不行，大佬试过将旷世的shufflenetv2+模型转ncnn吗，onnx-simplifier会报错

**Comment by nihui** [comment]
针对onnx模型转换的各种问题，推荐使用最新的pnnx工具转换到ncnn
In view of various problems in onnx model conversion, it is recommended to use the latest pnnx tool to convert your model to ncnn

```shell
pip install pnnx
pnnx model.onnx inputshape=[1,3,224,224]
```

详细参考文档
Detailed reference documentation
https://github.com/pnnx/pnnx
https://github.com/Tencent/ncnn/wiki/use-ncnn-with-pytorch-or-onnx#how-to-use-pnnx
